toolkit/tools/imagegen/configuration/networkconfig.go

```diff
@@ -0,0 +1,309 @@
+	NameServers []string `json:"NameServer"`
+	Device      string   `json:"Device"`
+}
+
+// valid network boot protocols supported
+var validBootProtos = map[string]bool{
+	"dhcp":   true,
+	"static": true,
+	"bootp":  true,
+	"ibft":   true,
+	"":       true,
+}
+
+// UnmarshalJSON Unmarshals a Network entry
+func (n *Network) UnmarshalJSON(b []byte) (err error) {
+	// Use an intermediate type which will use the default JSON unmarshal implementation
+	type IntermediateTypeNetwork Network
+	err = json.Unmarshal(b, (*IntermediateTypeNetwork)(n))
+	if err != nil {
+		return fmt.Errorf("failed to parse [Network]: %w", err)
+	}
+
+	// Now validate the resulting unmarshaled object
+	err = n.IsValid()
+	if err != nil {
+		return fmt.Errorf("failed to parse [Network]: %w", err)
+	}
+	return
+}
+
+// IsValid returns an error if the Network struct is not valid
+func (n *Network) IsValid() (err error) {
+	err = n.bootProtoIsValid()
+	if err != nil {
+		return
+	}
+
+	err = n.ipAddrIsValid()
+	if err != nil {
+		return
+	}
+
+	err = n.deviceIsValid()
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// ConfigureNetwork performs network configuration during the kickstart unattended installation process
+func ConfigureNetwork(installChroot *safechroot.Chroot, systemConfig SystemConfig) (err error) {
+	const squashErrors = false
+	var dnsUpdate bool
+
+	for _, networkData := range systemConfig.Networks {
+		deviceName, err := checkNetworkDeviceAvailability(networkData)
+		if err != nil {
+			return err
+		} else if deviceName == "" {
+			err = fmt.Errorf("The input network device is currently not supported on this system")
+			return err
+		}
+
+		err = createNetworkConfigFile(installChroot, networkData, deviceName)
+		if err != nil {
+			return err
+		}
+
+		dnsUpdate = len(networkData.NameServers) > 0
+	}
+
+	err = shell.ExecuteLive(squashErrors, "systemctl", "restart", "systemd-networkd")
+	if err != nil {
+		logger.Log.Errorf("Unable to restart systemd-networkd: %s", err)
+		return
+	}
+
+	if dnsUpdate {
+		err = shell.ExecuteLive(squashErrors, "systemctl", "restart", "systemd-resolved")
+		if err != nil {
+			logger.Log.Errorf("Unable to restart systemd-resolved: %s", err)
+		}
+	}
+
+	return
+}
+
+// bootProtoIsValid returns an error if input bootproto is invalid
+func (n *Network) bootProtoIsValid() (err error) {
+	n.BootProto = strings.TrimSpace(n.BootProto)
+	if validBootProtos[n.BootProto] {
+		return
+	}
+
+	return fmt.Errorf("invalid input for bootproto (%s), bootproto can only be one of dhcp, bootp, ibft and static", n.BootProto)
+}
+
+func (n *Network) validateIPAddress(ip string) (err error) {
+	if ip == "" && (n.BootProto == "" || n.BootProto == "dhcp") {
+		return
+	} else if net.ParseIP(ip) == nil {
+		return fmt.Errorf("invalid ip address (%s)", ip)
+	}
+
+	return
+}
+
+// ipAddrIsValid returns an error if ip, gateway, netmask or nameserver inputs are invalid ip addresses
+func (n *Network) ipAddrIsValid() (err error) {
```

Comment:
Nit: considering what's happening in the function, plural might be better. For instance: `ipAddressesAreValid`.